tectonic, tectonic-unwrapped, texpresso: fix build & clean up

bryango opened the pull request:
Supersedes:
- #384610
- fix build for linux, use `clangStdenv`
- clean up for darwin

See #384610 for context.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

doronbehar committed:
tectonic: cleanup Darwin specific apple_sdk usage

bryango committed:
tectonic, tectonic-unwrapped: fix linux build

bryango commented:
Weird, I don't have this eval error locally... Probably something from master that hasn't gone into my local branch. Let me see...

doronbehar commented:
I'm working on that.

bryango committed:
texpresso, texpresso.tectonic: simplify overrides

The previous implementation (overriding `buildRustPackage`) breaks
downstream `buildRustPackage.overrides` in subtle ways. It's thus better
to use `.overrideAttrs` explicitly and reconstruct `cargoDeps` when
needed.

This commit should cause no rebuild, but the underlying change is
necessary for an upcoming patch on tectonic itself.

bryango commented:
I rewrote parts of the `texpresso.tectonic` overrides and it seems to work now. `texpresso` should not be rebuilt by these changes; only the overriding logic is refactored.

doronbehar commented on pkgs/tools/typesetting/tex/texpresso/tectonic.nix:
```diff
@@ -16,18 +10,23 @@ tectonic-unwrapped.override (old: {
           };
           useFetchCargoVendor = true;
           cargoHash = "sha256-mqhbIv5r/5EDRDfP2BymXv9se2NCKxzRGqNqwqbD9A0=";
+          # rebuild cargoDeps by hand because `.overrideAttrs cargoHash`
+          # does not reconstruct cargoDeps (a known limitation):
+          cargoDeps = rustPlatform.fetchCargoVendor {
+            inherit (finalAttrs) src;
+            name = "${finalAttrs.pname}-${finalAttrs.version}";
+            hash = finalAttrs.cargoHash;
+            patches = finalAttrs.cargoPatches;
+          };
```
Well done! I was very close but my internet connection was too slow to try out things in a reasonable manner :).

doronbehar reviewed (commented):
So is this ready? Or not well tested? Can we mark it as ready?

bryango commented:
> So is this ready? Or not well tested? Can we mark it as ready?

I haven't tested the build yet (only tested eval); but if it builds fine (and `tectonic.passthru.tests` succeed) then it should be ready!

doronbehar commented:
Also tested the `tectonic.passthru.tests` derivation to work.

doronbehar committed:
texpresso.tectonic: fix meta

doronbehar committed:
texpresso.tectonic: nixfmt

doronbehar committed:
texpresso.tectonic: set version

doronbehar committed:
texpresso: callPackage ./tectonic.nix from package expression

doronbehar committed:
texpresso: move to pkgs/by-name

doronbehar committed:
texpresso: modernize / standardize

Move src closer to version attribute, so diffs of version bumps will be
smaller and easier to read, and a version attribute changing without a
hash changing will be more apparent, in case that happens.

bryango commented:
- built on 3 platforms: https://github.com/bryango/nix-build-action/actions/runs/13499401200
- ran a tectonic compile on my own .tex on `x86_64-linux`

I think it's good to go! Do we need to ping texpresso's maintainers? Or maybe we merge first since it's fixing broken builds?

doronbehar commented:
> I think it's good to go! Do we need to ping texpresso's maintainers?

I wonder why CI hasn't pinged them already - there are many commits here prefixed with `texpresso`. But anyway, cc @nickhu .

doronbehar commented:
Let's give him say a day. Some of the changes here are somewhat stylistic and perhaps debatable.

doronbehar commented:
I suspect that it's due to the draft state that was modified during commits pushed, but I won't put an effort to verify & debug this.

MithicSpirit commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 384706`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>tectonic</li>
    <li>tectonic-unwrapped</li>
    <li>texpresso</li>
  </ul>
</details>
